gossip: avoid removing nodes that get a new address

knz opened the pull request:
Fixes #34120.

K8s deployments make it possible for a node to get restarted using an
address previously attributed to another node, *while the other node
is still alive* (for example, a re-shuffling of node addresses during
a rolling restart).

Prior to this patch, the gossip code was assuming that if a node
starts with an address previously attributed to another node, that
other node must be dead, and thus was (incorrectly) *erasing* that
node's entry, thereby removing it from the cluster.

This scenario can be reproduced like this:

- start 4 nodes n1-n4
- stop n3 and n4
- restart n3 with n4's address

Prior to this patch, this scenario would yield "n4 removed from the
cluster" in other nodes, and n3 was not restarting properly. With the
patch, there is a period of time (until
`server.time_until_store_dead`) during which Raft is confused to not
find n4 at n3's address, but where the cluster otherwise operates
normally. After the store time outs, n4 is properly marked as down and
the log spam stops.

Release note (bug fix): CockroachDB now enables re-starting a node at
an address previously allocated for another node.

knz committed:
gossip: avoid removing nodes that get a new address

K8s deployments make it possible for a node to get restarted using an
address previously attributed to another node, *while the other node
is still alive* (for example, a re-shuffling of node addresses during
a rolling restart).

Prior to this patch, the gossip code was assuming that if a node
starts with an address previously attributed to another node, that
other node must be dead, and thus was (incorrectly) *erasing* that
node's entry, thereby removing it from the cluster.

This scenario can be reproduced like this:

- start 4 nodes n1-n4
- stop n3 and n4
- restart n3 with n4's address

Prior to this patch, this scenario would yield "n4 removed from the
cluster" in other nodes, and n3 was not restarting properly. With the
patch, there is a period of time (until
`server.time_until_store_dead`) during which Raft is confused to not
find n4 at n3's address, but where the cluster otherwise operates
normally. After the store time outs, n4 is properly marked as down and
the log spam stops.

Release note (bug fix): CockroachDB now enables re-starting a node at
an address previously allocated for another node.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34155)
<!-- Reviewable:end -->

knz commented:
@petermattis can you hand hold me about where to implement a test for this? Should this be a unit test, a roachtest or something else entirely? (cli acceptance test perhaps?)

Where should that test live? (Which directory?)

here's the test script I'm working with, for reference:

```shell
#! /bin/sh

C=$PWD/cockroach
d=$PWD/repro-34120

killall -9 cockroach cockroachshort

rm -rf "$d"
mkdir -p "$d"
cd "$d"

export COCKROACH_SCAN_MAX_IDLE_TIME=20ms

# Start cluster
for n in 1 2 3 4; do
    (
        mkdir -p d$n
        cd d$n
        $C start --background --listen-addr=localhost:2600$n --http-addr=localhost:2700$n --insecure --join=localhost:26001
    )
    if test $n = 1; then
        # Init cluster
        $C init --host localhost:26001 --insecure
    fi
    # We sleep to ensure node IDs are generated in the same order as directories.
    sleep 1
done

# Make store death detected faster.
$C sql --host localhost:26001 --insecure --echo-sql -e "SET CLUSTER SETTING server.time_until_store_dead = '1m15s'"

# Create some ranges
(
    for i in `seq 1 10`; do
        echo "create table t$i(x int); insert into t$i(x) values(1);"
    done
) | $C sql --insecure --host localhost:26001

# Wait for cluster to stabilize
tail -f d1/cockroach-data/logs/cockroach.log &
tailpid=$!
while true; do
    sleep 2
    lastlines=$(tail -n 2 d1/cockroach-data/logs/cockroach.log | grep 'runtime stats' | wc -l)
    if test $lastlines -lt 2; then
        continue
    fi
    break
done
kill -9 $tailpid
echo "OK"

# Stop the last two nodes.
$C quit --host localhost:26003 --insecure
$C quit --host localhost:26004 --insecure

# Restart with node shift
(
    cd d3
    $C start --background --listen-addr=localhost:26004 --http-addr=localhost:27004 --insecure
)

# Start a load on all tables
(
    echo '\set show_times on'
    while true; do
        sleep 1
        for i in `seq 1 10`; do
            echo "select * from t$i limit 1;"
        done
    done
) | $C sql --insecure --echo-sql --host localhost:26001 &>sql.log &

exec tail -f d1/cockroach-data/logs/cockroach.log
```

knz commented:
(Note that I'm waiting on CI to tell me which other tests need adjusting. I'll do this adjustment myself)

knz commented:
I'm a bit puzzled about why raft is confused. I'd expect the restarted node to ensure the other nodes have an updated address for it.

petermattis commented:
> I'm a bit puzzled about why raft is confused. I'd expect the restarted node to ensure the other nodes have an updated address for it.

Perhaps the store-not-found error is not causing `storage/raft_transport.go` to create a new connection. Note that currently we don't actually do any check that the receiving node is the correct node, even though we do have that information. See `RaftTransport.handleRaftRequest`.

petermattis commented:
PS I connected with @knz privately on testing.

petermattis commented:
Hmm, there is code in `Store.HandleRaftResponse` that claims we do close down the Raft connection when receiving a `StoreNotFoundError`. Perhaps that isn't working correctly.

tbg commented on pkg/gossip/gossip.go:
```diff
@@ -840,21 +840,8 @@ func (g *Gossip) updateNodeAddress(key string, content roachpb.Value) {
 		log.Infof(ctx, "removing n%d which was at same address (%s) as new node %v",
 			oldNodeID, desc.Address, desc)
 		g.removeNodeDescriptorLocked(oldNodeID)
```
Why isn't this also gone?

knz replied:
I figured that that gossip entry is known to be stale at that point so it may as well be removed from gossip until a fresh one is received.

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34155)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis and @tbg)

___
*[pkg/gossip/gossip.go, line 842 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWlYtvGZRSdaSHz0cA5-r1-842:-LWmMdpe-sKluiMAgVVK:b-jhbvbk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fdc4a869d8e7dc2bfbc5a45cc746a174a421aa9/pkg/gossip/gossip.go#L842)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I figured that that gossip entry is known to be stale at that point so it may as well be removed from gossip until a fresh one is received.
</blockquote></details>

 This isn't removing the gossip entry, but the entry from `Gossip.nodeDescs`. But that entry might be up to date. Consider the scenario where node 1 and node 2 swap addresses. Node 1 might start up and gossip its new address. This will update the `Gossip.nodeDescs` map on every receiving node, and remove the entry for node 2. When node 2 gossips its new address (node 1's old address), this line will remove the `nodeDescs` entry for node 1 which is not correct. We could fix this by only removing the `nodeDescs` entry if the address matches what is in `bootstrapAddrs`. 


<!-- Sent from Reviewable.io -->

knz commented:
Thanks for explaining.

knz commented:
I have modified the PR to make the tests pass. That may be sufficient for the customer who raised the issue.
However I am also working to resolve #34158 to make the fix more ironclad.

tbg commented:
@petermattis I also re-analyzed why this code was introduced in the first place in https://github.com/cockroachdb/cockroach/issues/34120#issuecomment-456365507. Seems to me that it was just a completely incorrect fix.

knz commented:
Reworked the PR; added a 2nd commit with the fix to #34158.
I have verified that it makes the warnings from the test scenario completely go away (raft finds itself happy nearly instantaneously).

The following warnings remain however, emitted on n1 when n4 is restarted at n3's address:

```
W190122 14:12:54.422039 885 vendor/google.golang.org/grpc/clientconn.go:1440  grpc: addrConn.transportMonitor exits due to: grpc: the connection is closing
W190122 14:12:54.442055 2372 vendor/google.golang.org/grpc/clientconn.go:1304  grpc: addrConn.createTransport failed to connect to {localhost:26003 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp [::1]:26003: connect: connection refused". Reconnecting...
I190122 14:12:54.442164 2358 vendor/github.com/cockroachdb/circuitbreaker/circuitbreaker.go:322  [n1] circuitbreaker: rpc 127.0.0.1:26001->3 tripped: failed to grpc dial n3 at localhost:26003: initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest
connection error: connection error: desc = "transport: Error while dialing dial tcp [::1]:26003: connect: connection refused"
I190122 14:12:54.442186 2358 vendor/github.com/cockroachdb/circuitbreaker/circuitbreaker.go:447  [n1] circuitbreaker: rpc 127.0.0.1:26001->3 event: BreakerTripped
I190122 14:12:54.442196 2358 rpc/nodedialer/nodedialer.go:95  [ct-client] unable to connect to n3: failed to grpc dial n3 at localhost:26003: initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transpo
rt: Error while dialing dial tcp [::1]:26003: connect: connection refused"
W190122 14:12:55.448289 2372 vendor/google.golang.org/grpc/clientconn.go:1304  grpc: addrConn.createTransport failed to connect to {localhost:26003 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
W190122 14:12:55.448312 2372 vendor/google.golang.org/grpc/clientconn.go:953  Failed to dial localhost:26003: grpc: the connection is closing; please retry.
W190122 14:12:55.482671 1289 vendor/google.golang.org/grpc/clientconn.go:1304  grpc: addrConn.createTransport failed to connect to {localhost:26004 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
W190122 14:12:55.482698 1295 storage/raft_transport.go:583  [n1] while processing outgoing Raft queue to node 4: rpc error: code = Unavailable desc = transport is closing:
W190122 14:12:55.482708 1289 vendor/google.golang.org/grpc/clientconn.go:1440  grpc: addrConn.transportMonitor exits due to: context canceled
W190122 14:12:55.648681 2377 vendor/google.golang.org/grpc/clientconn.go:1304  grpc: addrConn.createTransport failed to connect to {localhost:26004 0  <nil>}. Err :connection error: desc = "transport: Error while dialing dial tcp [::1]:26004: connect: connection refused". Reconnecting...
I190122 14:12:55.648764 2314 vendor/github.com/cockroachdb/circuitbreaker/circuitbreaker.go:322  [n1] circuitbreaker: rpc 127.0.0.1:26001->4 tripped: failed to grpc dial n4 at localhost:26004: initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest
connection error: connection error: desc = "transport: Error while dialing dial tcp [::1]:26004: connect: connection refused"
I190122 14:12:55.648777 2314 vendor/github.com/cockroachdb/circuitbreaker/circuitbreaker.go:447  [n1] circuitbreaker: rpc 127.0.0.1:26001->4 event: BreakerTripped
I190122 14:12:55.648784 2314 rpc/nodedialer/nodedialer.go:95  [ct-client] unable to connect to n4: failed to grpc dial n4 at localhost:26004: initial connection heartbeat failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transpo
rt: Error while dialing dial tcp [::1]:26004: connect: connection refused"
W190122 14:12:56.655188 2377 vendor/google.golang.org/grpc/clientconn.go:1304  grpc: addrConn.createTransport failed to connect to {localhost:26004 0  <nil>}. Err :connection error: desc = "transport: Error while dialing cannot reuse client connection". Reconnecting...
W190122 14:12:56.655201 2377 vendor/google.golang.org/grpc/clientconn.go:953  Failed to dial localhost:26004: grpc: the connection is closing; please retry.
I190122 14:12:56.720259 105 vendor/github.com/cockroachdb/circuitbreaker/circuitbreaker.go:447  [n1] circuitbreaker: rpc 127.0.0.1:26001->3 event: BreakerReset
```

I'd like to make these warnings go away, **or** become more descriptive. How come the error I added in the code ("client requested node ID doesn't match") is not printed by grpc? Does it not get transmitted back somehow?

knz commented:
Ok so except for the missing roachtest/roachprod harness, which may take me a day or two, I think the bug fix is ready!

(Note: the PR does have unit tests at the level of the RPC dial logic and gossip. Perhaps we can decouple the bug fix and the more comprehensive acceptance test?).

tbg reviewed (commented):
The GRPC logging is a mess. I think you should ignore it for now, but we should consider bumping it to ERROR:

https://github.com/cockroachdb/cockroach/blob/d591a49931a665f896b064363eddd975281ab197/pkg/cli/start.go#L447

Filed https://github.com/cockroachdb/cockroach/issues/34165. I think that for here, this leaves the circuitbreaker events, which again I would massage separately (taking a node down today yields the same errors).
I would change this line: https://github.com/cockroachdb/cockroach/blob/8636a0c4e3e17aa76b12086ddcee15a3b4468f5d/pkg/rpc/nodedialer/nodedialer.go#L206 so that it explicitly says the nodeid (`%v [n%d]` instead of `%v -> %v` which is just confusing). The breaker should log only once per minute, so I am confused why you're seeing it log much more frequently.

I think you don't need a migration for the changes to PingRequest since the field is optional and verified by the recipient. I.e. the field will be dropped silently, which works fine.

Other than these remarks the PR looks pretty good to me!

> Reviewed 2 of 2 files at r2, 16 of 16 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-:-LWq1v3H72LNIoVL0Lxp:bg78t8c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/gossip/gossip_test.go, line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWq-1oc7VK-fJZ08t6G:-LWq-1oc7VK-fJZ08t6H:b-v5rupx) ([raw file](https://github.com/cockroachdb/cockroach/blob/321a146fbe488d95a137d37be24c0d0ec6584bd4/pkg/gossip/gossip_test.go#L64)):*
> ```Go
> 
> // TestGossipMoveNode verifies that if a node is moved to a new address, it
> // gets properly updated in gossip (including that any other node that was
> ```

Stale comment.
___
*[pkg/rpc/context.go, line 668 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWq-XoX50pfbkFMVlSN:-LWq-XoX50pfbkFMVlSO:b4gar9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/111c31b1b2474d46b816c79d8c11360b7cd1db78/pkg/rpc/context.go#L668)):*
> remain
___
*[pkg/rpc/context.go, line 674 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWq-_1f51K2XJ6dj8XE:-LWq-_1f51K2XJ6dj8XF:b-fnc18o) ([raw file](https://github.com/cockroachdb/cockroach/blob/111c31b1b2474d46b816c79d8c11360b7cd1db78/pkg/rpc/context.go#L674)):*
> GRPDDialNode
___
*[pkg/rpc/context.go, line 788 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWq-tvg2Isuj06_2Jpz:-LWq-tvg2Isuj06_2Jq-:blh3y5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/111c31b1b2474d46b816c79d8c11360b7cd1db78/pkg/rpc/context.go#L788)):*
> ```Go
> 			MaxOffsetNanos: maxOffsetNanos,
> 			ClusterID:      &clusterID,
> 			NodeID:         conn.remoteNodeID,
> ```

`remoteNodeID` won't change out from under us, though? The comment above suggests that.
___
*[pkg/rpc/context_test.go, line 1167 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWq0Dsb1gApltJmawIu:-LWq0Dsb1gApltJmawIv:b-f9qbaw) ([raw file](https://github.com/cockroachdb/cockroach/blob/111c31b1b2474d46b816c79d8c11360b7cd1db78/pkg/rpc/context_test.go#L1167)):*
> ```Go
> func BenchmarkGRPCDialNode(b *testing.B) {
> 	if testing.Short() {
> 		b.Skip("TODO: fix benchmark")
> ```

Reminder to self


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34155)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/gossip/gossip_test.go, line 64 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWq-1oc7VK-fJZ08t6G:-LWqAW1c3o5Gxgy3Cf57:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/321a146fbe488d95a137d37be24c0d0ec6584bd4/pkg/gossip/gossip_test.go#L64)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Stale comment.
</blockquote></details>

Done.
___
*[pkg/rpc/context.go, line 668 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWq-XoX50pfbkFMVlSN:-LWqAYJd9-dLQ63bKE5Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/111c31b1b2474d46b816c79d8c11360b7cd1db78/pkg/rpc/context.go#L668)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

> remain
</blockquote></details>

Done.
___
*[pkg/rpc/context.go, line 674 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWq-_1f51K2XJ6dj8XE:-LWqAYkr2F_B_WOr5u6C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/111c31b1b2474d46b816c79d8c11360b7cd1db78/pkg/rpc/context.go#L674)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

> GRPDDialNode
</blockquote></details>

Done.
___
*[pkg/rpc/context.go, line 788 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWq-tvg2Isuj06_2Jpz:-LWqAgQr-eYj8h_P7zh1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/111c31b1b2474d46b816c79d8c11360b7cd1db78/pkg/rpc/context.go#L788)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

`remoteNodeID` won't change out from under us, though? The comment above suggests that.
</blockquote></details>

Done.
___
*[pkg/rpc/context_test.go, line 1167 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWq0Dsb1gApltJmawIu:-LWqAhP_CikSIXHuYuee:banlyuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/111c31b1b2474d46b816c79d8c11360b7cd1db78/pkg/rpc/context_test.go#L1167)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Reminder to self
</blockquote></details>

I copy/pasted the benchmark from above; is there anything I should change in the TODO?


<!-- Sent from Reviewable.io -->

knz commented:
> I would change this line:
> cockroach/pkg/rpc/nodedialer/nodedialer.go
> so that it explicitly says the nodeid (%v [n%d] instead of %v -> %v which is just confusing).

Also done.

knz commented:
Out of curiosity: there's about 1 minute between the moment the node is restarted at its new address, and the moment grpc dials into it. Where is that 1 minute delay?

(The 3 seconds heartbeat ensures that the connection to both of the nodes that are shut down are dropped quickly; however there must be another delay somewhere to pick up the restarted node. I don't know what that delay is.)

knz commented:
Oh I think that's the "time until store dead" (1m15s).

Which raises a follow up question: why is the "time until store dead" delay used in this case?

The node is being restarted (albeit at a different address) and thus is definitely not dead. Why is it not being reconnected to immediately when the original connection to it is dropped?

tbg reviewed (commented):
Let's wait for @petermattis to give this a thorough look, but it looks good to me.

> Also done.

I don't see it, though I might've missed it.

> Out of curiosity: there's about 1 minute between the moment the node is restarted at its new address, and the moment grpc dials into it. Where is that 1 minute delay?

What do you mean? Are you perhaps confused by the circuit breaker logging here? https://github.com/cockroachdb/cockroach/blob/8636a0c4e3e17aa76b12086ddcee15a3b4468f5d/pkg/rpc/nodedialer/nodedialer.go#L38

A node should be contacted as soon as it's available (with maybe a few sec of delay).

Writing the roachtest will be a little awkward since they're really not made for this sort of thing (moving store directories around). I have wondered before whether some tests should just deploy a single-node bash script, but that will spiral out of control and not in a good way. @petermattis how do you think we should handle this? See https://github.com/cockroachdb/cockroach/issues/33120 for another thing worth automating that would be awkward to roachtestify.

> Reviewed 1 of 16 files at r3, 3 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-:-LWqBdgK2wiMpIR4ZmZp:bni08a8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/rpc/context_test.go, line 1167 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWq0Dsb1gApltJmawIu:-LWqBUra8TUYK1bQr_iK:b-3wakrj) ([raw file](https://github.com/cockroachdb/cockroach/blob/111c31b1b2474d46b816c79d8c11360b7cd1db78/pkg/rpc/context_test.go#L1167)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I copy/pasted the benchmark from above; is there anything I should change in the TODO?
</blockquote></details>

I would remove the copy you introduced.


<!-- Sent from Reviewable.io -->

knz commented:
That's not what I see. Instead I see this:

1. time 0: node n3 gets restarted at n4's previous address
2. time 0 + epsilon: grpc errors in existing connections from n1 and n2 to n3/n4 - this is expected
3. time + ~1m: n1 receives error "intended node ID 4 does not match actual node ID 3"  - n1 has tried to connect to n4 (the dead node) to its old address, now taken by n3.
4. time + ~1m: n1 establishes starts streaming snapshots to n3 - this seems independent from the previous step

Questions:

- I don't care about the delay between step 2 and 3, although I don't understand it. I had expected a reconnect to n4 (now dead) to happen immediately after the connection wast lost at step 2.
- The delay between step 2 and 4 bothers me. I'd have expected n1 to establish a connection to n3 (restarted, fine albeit at new address) within seconds, as soon as an updated address was received (which should be within seconds, thanks to gossip)

knz reviewed (commented):
> I don't see it, though I might've missed it.

forgot to push

> What do you mean? Are you perhaps confused by the circuit breaker logging here? 

see my previous comment above

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-:-LWrdi3W8BgGyGqjJsra:b671yjz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/rpc/context_test.go, line 1167 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWq0Dsb1gApltJmawIu:-LWrdgBo1cvMmy5p2mN8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/111c31b1b2474d46b816c79d8c11360b7cd1db78/pkg/rpc/context_test.go#L1167)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I would remove the copy you introduced.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> Writing the roachtest will be a little awkward since they're really not made for this sort of thing (moving store directories around). I have wondered before whether some tests should just deploy a single-node bash script, but that will spiral out of control and not in a good way. @petermattis how do you think we should handle this?

My suggestion is to just move the store directories around. `roachprod put` already does scp between remote nodes in `treedist` mode. There might be a little plumbing, but it should be possible to `roachprod run <cluster> -- scp <src> <dest>`. Or perhaps this would be better as `roachprod cp <cluster> <node>:<src> <node>:<dest>`. Regardless, I think this can be done in a follow-on PR. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-:-LWsOxbM64QqRS97hW-z:b-j3wae6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @tbg)

___
*[pkg/rpc/context.go, line 671 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWsNhi18BQntGTA-moM:-LWsNhi18BQntGTA-moN:b-c16nds) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5042aae0b63e04d4c9718685e5163668a887137/pkg/rpc/context.go#L671)):*
> ```Go
> // a new one is created without a node ID requirement.
> func (ctx *Context) GRPCDial(target string) *Connection {
> 	return ctx.GRPCDialNode(target, 0)
> ```

So gossip will use a different connection that non-gossip? I think this effectively doubles the number of internal network connections. Is that a problem? 

Rather that adding the `remoteNodeID` to the map key, I was imagining that the `Connection` would contain a `remoteNodeID` atomic that is populated by the first `PingResponse` containing a `NodeID`. 
___
*[pkg/rpc/context.go, line 680 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWsQ1gU9WIyphnj4dhs:-LWsQ1gU9WIyphnj4dht:b-1428c9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5042aae0b63e04d4c9718685e5163668a887137/pkg/rpc/context.go#L680)):*
> ```Go
> // a connection already existed with a different node ID requirement
> // it is first dropped and a new one with the proper node ID
> // requirement is created.
> ```

This last sentence doesn't match the code, unless I'm misunderstanding the code. It seems like an existing connection under a different remote node ID is left untouched.
___
*[pkg/rpc/heartbeat.go, line 99 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWsO6Jw4rSaTH1A-ih1:-LWsO6Jw4rSaTH1A-ih2:b-rj5dqo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5042aae0b63e04d4c9718685e5163668a887137/pkg/rpc/heartbeat.go#L99)):*
> ```Go
> 		nodeID = hs.nodeID.Get()
> 	}
> 	if args.NodeID != 0 && nodeID != 0 && args.NodeID != nodeID {
> ```

Shouldn't it be an error if the local node doesn't have an ID yet (`nodeID == 0`) while the remote node requested a specific ID? I think this check should be `args.NodeID != 0 && args.NodeID != nodeID`.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34155)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @petermattis, and @tbg)

___
*[pkg/rpc/context.go, line 671 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWsNhi18BQntGTA-moM:-LWu88Bk0Q2JMoeMtIrP:bliuuu9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5042aae0b63e04d4c9718685e5163668a887137/pkg/rpc/context.go#L671)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

So gossip will use a different connection that non-gossip? I think this effectively doubles the number of internal network connections. Is that a problem? 

Rather that adding the `remoteNodeID` to the map key, I was imagining that the `Connection` would contain a `remoteNodeID` atomic that is populated by the first `PingResponse` containing a `NodeID`. 
</blockquote></details>

Ok I'll try that. 
___
*[pkg/rpc/heartbeat.go, line 99 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWsO6Jw4rSaTH1A-ih1:-LWu8L7XEc_NYevVNEE9:bghtjpc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5042aae0b63e04d4c9718685e5163668a887137/pkg/rpc/heartbeat.go#L99)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Shouldn't it be an error if the local node doesn't have an ID yet (`nodeID == 0`) while the remote node requested a specific ID? I think this check should be `args.NodeID != 0 && args.NodeID != nodeID`.
</blockquote></details>

I have found (at least in some tests) that there are places where a HeartbeatService is instantiated without a valid node ID and I wasn't sure about that, hence this conditional. I'll change it back and report with details.


<!-- Sent from Reviewable.io -->

ajwerner commented:
>The breaker should log only once per minute, so I am confused why you're seeing it log much more frequently.

I added new logging inside the circuit breaker itself. That new logging does not use the log.Every, though maybe it should. To use it we'd need to modify the calls [here](https://github.com/cockroachdb/cockroach/blob/f37d09cd3cdd32f4d4894611cfd60caf25c10fff/pkg/rpc/breaker.go#L104)

When adding that logging I wasn't overly concerned about generating too much noise. Perhaps I should have been.

knz commented:
I'm going to fork the RPC change to a separate PR, so that the first change can be merged ahead of full resolution.

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34155)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @petermattis, and @tbg)

___
*[pkg/rpc/context.go, line 696 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-LWw-zedDV8AG9Vlfsh9:-LWw-zedDV8AG9VlfshA:b-4cia0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/33afd7f609b0b97a47b1a2896bc48104aab2af86/pkg/rpc/context.go#L696)):*
> ```Go
> 			// ensure we're registering the connection we just created for
> 			// future use by these other dials.
> 			_, _ = ctx.conns.LoadOrStore(connKey{target, 0}, value)
> ```

One problem with this approach is that we won't remove this conn if there is a heartbeat or RPC error: we'll only remove the entry with the non-zero `remoteNodeID`. I think that can be fixed, but I'm not sure the complexity is worth it. I know I led you down this path, but thinking about it more I think I was in error. One mistake in my earlier thinking is that the number of gossip connections per-node is limited. So we won't be doubling the number of connections, and even when we do we're talking about a relatively small number of connections in the cluster until we reach very large cluster sizes.

I'd prefer to just remove this block of code and let there between 2 connections per node.


<!-- Sent from Reviewable.io -->

knz commented:
Done - I have rewinded this PR to just the first commit, the other is #34197.

Without the other PR #34197 we keep the log spam, because of the reasons explained in that PR.

```
W190123 18:43:46.037282 318 storage/raft_transport.go:583  [n1] while processing outgoing Raft queue to node 4: store 4 was not found:
W190123 18:43:46.040255 3813 storage/store.go:3744  [n1,s1,r9/1:/Table/1{3-4}] raft error: node 4 claims to not contain store 4 for replica (n4,s4):4: store 4 was not found
W190123 18:43:46.040267 3826 storage/raft_transport.go:583  [n1] while processing outgoing Raft queue to node 4: store 4 was not found:
W190123 18:43:46.053504 3828 storage/store.go:3744  [n1,s1,r9/1:/Table/1{3-4}] raft error: node 4 claims to not contain store 4 for replica (n4,s4):4: store 4 was not found
W190123 18:43:46.053517 3800 storage/raft_transport.go:583  [n1] while processing outgoing Raft queue to node 4: store 4 was not found:
W190123 18:43:46.058747 3830 storage/store.go:3744  [n1,s1,r9/1:/Table/1{3-4}] raft error: node 4 claims to not contain store 4 for replica (n4,s4):4: store 4 was not found
```

knz commented:
> One  problem with this approach is ...
> I'd prefer to just remove this block of code and let there between 2 connections per node.

Will do in the other PR

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34155#-:-LWw3VPK8ICo0wwErVS1:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz, @petermattis, and @tbg)



<!-- Sent from Reviewable.io -->

knz commented:
Ok then we'll defer the larger test work to a separate PR.

bors r+
